ext/span.c

```diff
@@ -796,3 +823,78 @@ zend_string *ddtrace_trace_id_as_hex_string(ddtrace_trace_id id) {
     snprintf(ZSTR_VAL(str), 33, "%016" PRIx64 "%016" PRIx64, id.high, id.low);
     return str;
 }
+
+void free_inferred_proxy_result(ddtrace_inferred_proxy_result *result) {
+    zend_string_release(result->system);
+    zend_string_release(result->start_time_ms);
+    if (result->http_method) zend_string_release(result->http_method);
+    if (result->path) zend_string_release(result->path);
+    if (result->domain) zend_string_release(result->domain);
+    if (result->stage) zend_string_release(result->stage);
+}
+
+
+ddtrace_root_span_data *ddtrace_open_inferred_span(zend_array *headers) {
+    ddtrace_read_header *read_header = headers ? ddtrace_read_array_header : ddtrace_read_zai_header;
+    ddtrace_inferred_proxy_result result = ddtrace_read_inferred_proxy_headers(read_header, headers);
+
+    if (!result.system || !result.start_time_ms) {
+        free_inferred_proxy_result(&result);
+        return NULL;
+    }
+
+    const ddtrace_proxy_info *proxy_info = ddtrace_get_proxy_info(result.system);
+    if (!proxy_info) {
+        zend_string_release(result.system);
+        zend_string_release(result.start_time_ms);
+        return NULL;
+    }
+
+    ddtrace_span_data *span = ddtrace_push_inferred_root_span();
+
+    zval_ptr_dtor(&span->property_name);
+    ZVAL_STR(&span->property_name, zend_string_init(proxy_info->span_name, strlen(proxy_info->span_name), 0));
+
+    zval_ptr_dtor(&span->property_resource);
+    if (result.http_method && result.path) {
+        ZVAL_STR(&span->property_resource, strpprintf(0, "%s %s", ZSTR_VAL(result.http_method), ZSTR_VAL(result.path)));
+    }
+
+    span->start = ZEND_ATOL(ZSTR_VAL(result.start_time_ms)) * 1000000;
+    span->duration_start = zend_hrtime() - (ddtrace_nanoseconds_realtime() - span->start);
+
+    zval zv;
+
+    if (result.domain) {
+        ZVAL_STR_COPY(&zv, result.domain);
+        ddtrace_assign_variable(&span->property_service, &zv);
```

Comment:
Is this `if (result.domain)` rather a safeguard or something which will regularly happen? Because if it can happen that the domain is not set, the inferred span will have the generic service name (http.request), while frameworks will have the service name from the root span. Which will certainly be unexpected.